### DISCUSSION TITLE
Creating HDR EXR files from 16 bit linear TIFF

### DISCUSSION THREAD

**Discussion by jbondhus** [ImageMagick/ImageMagick#5823] [Help]
Is it possible to make an EXR file that can be viewed in HDR from a 16 bit linear TIFF? Basically, I want to be able to take a wide gamut TIFF file (ideally using the ProPhotoRGB primaries in the TIFF file) and create an HDR image that can be viewed in an HDR image viewer on an HDR display. Right now my process involves dumping an under/over exposed version out using photoshop and then ingesting all three into Photomatix and exporting the intermediate result (which is an EXR image file). That's rather cumbersome and there has to be a better way. I can convert it right now using Image Magick but there's no metadata in it and imaging programs don't recognize it as an HDR image, it's basically just an SDR image wrapped in the EXR format.

**Comment by snibgo** [comment]
I don't have an HDR monitor, and perhaps I don't have an HDR viewer, so the following is untested.

> Is it possible to make an EXR file that can be viewed in HDR from a 16 bit linear TIFF?

16-bit integer? I doubt it. I suspect you need floating-point source. TIFF typically records floating-point values betwen 0.0 and 1.0, and I suspect HDR values extend above that, eg to 4.0 or 16.0 or whatever.

Of course, provided your IM has HDRI, you can simply multiply values (EDIT: from a linear source), like this:
```
magick in.tiff -evaluate Multiply 16 out.exr
```

>  ... but there's no metadata in it and imaging programs don't recognize it as an HDR image, ...

What metadata is missing? Perhaps you can insert that metadata with exiftool. If IM is missing something that makes proper HDR files, I expect that can be fixed.

**Comment by jbondhus** [comment]
EXR files support both 16 bit integer and 32 bit float, really it's just linear light levels. I thought a 16 bit color depth tiff would be 16 bit integers, wouldn't it be?

Anyways, your suggestion seems to work preliminarily, when I get back to my HDR monitor I'll take a look on that and see if it looks right.

![image](https://user-images.githubusercontent.com/1085350/204388317-2e39f3ab-c805-4243-b906-3423c66d9653.png)

**Comment by jbondhus** [comment]
The EXR conversion results in a file that's very flat, maybe gamma conversion of some sort messing it up. Outputting to a .hdr image file (radiance RGBE) the luminance and gamma is correct, however ImageMagick seems to be not copying over the color profile. I'm only able to get the colors looking right when I output 16 bit TIFF files in the sRGB color space, whereas both formats should support embedded color profiles.

**Comment by snibgo** [comment]
I confirm that IM doesn't write ICC colour profiles to EXR files. There seems to be nothing in coders\exr.c that tries to write profiles.

There seems to be a good reason for this: EXR doesn't support ICC colour profiles. See https://forum.affinity.serif.com/index.php?/topic/93544-saving-color-space-in-exr-32-bit-hdr-files/ and https://ninedegreesbelow.com/photography/embedded-color-space-information.html .

This surprises me, and seems weird. Software that reads an EXR file needs to know the colorimetric meaning of pixels values. This would come from metadata that defines the chromaticity of each primary, and any transfer response curve (TRC). ICC color profiles are a common mechanism for this. I suppose EXR files are always linear, so there is no TRC.

IM can set primaries, for example:
```
magick toes.png -profile sRGB.icc -profile Rec2020-elle-v4-g10.icc -blue-primary 0.3,0.6 x.exr
```
... but setting primaries makes no change to the output x.exr. This may be a deficiency in ImageMagick.

**Comment by jbondhus** [comment]
First, speaking of the lack of a transfer curve, the output of a tiff to EXR almost looks like it has a gamma curve applied. It's extremely low contrast, wouldn't that happen if a gamma curve was applied to something that's then output into a linear format? Or do I have that backwards?

Second, what about the .hdr file format, RGBE? That works correct at least for luminance and gamma as far as I can tell. That uses the XYZ color model so there's no color space of any kind to worry about. Is it possible that ImageMagick is assuming the input is sRGB when it does the XYZ conversion? Just spitballing here - I might try explicitly changing primaries in the RGBE format.

Also, since the Y, luminance, is dimensionless, how is peak luminance specified for the RGBE format? In my program it comes up as "peak brightness in nits (cd/m^2)", so I assume there's a place to set peak luminance and the luminance of each pixel is thus relative to that.

**Comment by jbondhus** [comment]
Nevermind about the second bit, the XYZ mode of the RGBE format is optional, so if you're using RGB you'll still have to write the color profile. I'll try manually setting the primaries and see if that solves it on the RGBE format at least.

**Comment by snibgo** [comment]
> First, speaking of the lack of a transfer curve, the output of a tiff to EXR almost looks like it has a gamma curve applied.

Is your input TIFF linear? If it is sRGB or Adobe1998 or similar with a gamma, you should make it linear first. This is because coders\exr.c *assumes* the input is linear, rather than automatically converting it to linear. Do this either with "`-colorspace RGB`" or with profiles:
```
magick in.tiff -colorspace RGB out.exr
```
... or ...
```
magick in.tiff -profile sRGB.icc -profile sRGB-elle-V4-g10.icc out.exr
```

The HDR format looks promising. coders\hdr.c doesn't *assume* the input is linear RGB, and will convert it if necessary. It can use image properties "hdr:exposure" (whatever that does) and primaries.

**Comment by jbondhus** [comment]
See attached for a sample file, in the ProPhotoRGB color space, 16 bit linear TIFF. I've tried outputting it in sRGB as well as ProPhoto RGB, the latter results in really messed up colors in the HDR format because the color profile doesn't seem to be honored. Both the sRGB output and the ProPhoto RGB linear file result in weird gamma issues when converting to EXR.

https://jbondhus-sharing.s3.us-east-1.amazonaws.com/2022/2022-11-29/_1001463.tif

Also, Exposure seems to be a multiplier on the original pixel values, see page 29 of the RGBE (Radiance HDR) file spec linked below. The way they describe it, it almost seems like the pixel values are absolute light levels. Primaries are also in the header, though it doesn't seem like ImageMagick is setting them, I tried overriding them as well to something way off and the image looked correct still.

https://floyd.lbl.gov/radiance/refer/filefmts.pdf

**Comment by jbondhus** [comment]
Also, I tried what you listed above, specifying the input color space when exporting to an EXR file - now the image doesn't look so flat. My understanding though is that the command flag "-colorspace RGB" forces imagemagick to assume the file is linear and **not** gamma corrected, is that wrong? The TIFF file should indeed be linear, feel free to take a look at the TIFF I uploaded.

**Comment by snibgo** [comment]
Your TIFF link gives "access denied".

> Primaries are also in the header, though it doesn't seem like ImageMagick is setting them, I tried overriding them as well to something way off and the image looked correct still.

What was your IM command?
```
magick -comment "by snibgo" toes.png -colorspace RGB -blue-primary 0.123,0.234 -set hdr:exposure 321 x0.hdr
```

The first few bytes of x0.hdr are text:
```
#?RADIANCE
#by snibgo
EXPOSURE=321
GAMMA=1
PRIMARIES=0 0 0 0 0.123 0.234 0 0
FORMAT=32-bit_rle_rgbe
```
I have set only the blue primary. Red, green and white should also be set.

> My understanding though is that the command flag "-colorspace RGB" forces imagemagick to assume the file is linear and not gamma corrected, is that wrong?

That is wrong. "-colorspace RGB" will, if the image is not already RGB, update the metadata and change all the pixel values. Contrast this with "-set colorspace RGB" which changes only metadata.

**Comment by jbondhus** [comment]
Oops, I moved it and it reset the permissions. Try again.

The command I tried last night was this. I think those primaries are correct for ProPhoto RGB, though I tried changing the blue primary significantly and didn't notice any change in the output image. I'l give your variant a shot.

```
magick.exe convert .\_1001532.tif -red-primary 0.734699,0.265301 -green-primary 0.159597,0.840403 -blue-primary 0.036598,0.000105 .\_1001532.hdr
```

Ah, didn't notice the different flag regarding colorspace vs set colorspace. Interesting that it fixed it, because that TIFF should be linear, unless Lightroom adds gamma correction to non sRGB color spaces when outputting a TIFF, which seems wrong...

**Comment by snibgo** [comment]
> See attached for a sample file, in the ProPhotoRGB color space, 16 bit linear TIFF.

> ... because that TIFF should be linear, ...

Why do you think it is linear?

The file _1001463.tif has an embedded profile, with a gamma about 1.8, the usual ProPhotoRGB gamma. 

To make it linear, either use "-evaluate Pow 1.8" or "-profile XX.icc" where XX.icc is some linear ProPhoto profile.

Elle Stone has published a linear ProPhoto profile, as "LargeRGB-elle-V4-g10.icc" in https://github.com/ellelstone/elles_icc_profiles/tree/master/profiles .

**Reply by fmw42** [reply]
@snibgo.   Just curious -- how does one tell which profiles are linear and which are not? I do not see a guide to the profiles. Are you reading the code? Or am I missing something else?

**Reply by snibgo** [reply]
@fmw42: IM told me there was an embedded profile. I viewed the file with two programs: one uses the profile and the other doesn't. The images had different colour saturations but similar intensities. This suggested the embedded profile was larger than sRGB but with a similar gamma.

I confirmed this with two programs: ICC profile inspector (from www.color.org) and iccDumpProfile (from I have forgotten where). These said the gamma was about 1.8.

**Reply by jbondhus** [reply]
So apparently lightroom writes 16 bit TIFF files with gamma correction, even if you're not using sRGB. I did not realize that.

That certainly explains the weirdness with the EXR image. Still the color gamut issue is not resolved, both the HDR and EXR formats don't appear to honor the color primaries when I manually specify them. If there's a workaround I'm fine with that too.

**Reply by fmw42** [reply]
I was asking about all of Elle Stone's profiles not what was in the file.

**Reply by jbondhus** [reply]
@fmw42 Oh yeah, I just wanted to make it clear that I was mistaken in my thought that lightroom exports TIFF files with no gamma correction when they mentioned that my file has a gamma of 1.8. Sorry for the confusion.

**Reply by snibgo** [reply]
@fmw42:

> I was asking about all of Elle Stone's profiles not what was in the file.

Sorry, I misunderstood your question.

The guide to Elle's profiles is at https://ninedegreesbelow.com/photography/lcms-make-icc-profiles.html#ProphotoRGB

Her profiles *-g10.icc are linear.

**Reply by fmw42** [reply]
Thanks.  That page is very informative regarding v2 and v4 profile.

**Comment by snibgo** [comment]
@jbondhus: You use "magick convert". This gives you the old v6 syntax, which I don't recommend unless you really *really* need v6. Instead, I suggest just "magick" (without "convert"), which gives the much better v7 syntax.

> ... both the HDR and EXR formats don't appear to honor the color primaries when I manually specify them.

What was your command? I have shown how to set the primaries in HDR. Whether an HDR reader honours them is a different matter.

**Comment by jbondhus** [comment]
@snibgo I'll use the v7 syntax, thanks.

```
magick.exe convert .\_1001532.tif -red-primary 0.734699,0.265301 -green-primary 0.159597,0.840403 -blue-primary 0.036598,0.000105 .\_1001532.hdr
```

The command I tried last night is listed above. Photoshop doesn't honor them either, it shows the color as way off on ProPhoto RGB images even setting the primaries explicitly, which is why I'm wondering if they're being set correctly. Since the image format header is plain text, I'll dig around some more and get back.

**Comment by snibgo** [comment]
You also need to set the https://imagemagick.org/script/command-line-options.php#white-point , otherwise that will be "0,0" which will probably give bad results.

Windows syntax:
```
magick.exe _1001463.tif ^
  -evaluate Pow 1.8 ^
  -set colorspace RGB ^
  -red-primary 0.734699,0.265301 ^
  -green-primary 0.159597,0.840403 ^
  -blue-primary 0.036598,0.000105 ^
  -white-point 0.345704,0.358540 ^
  x3.hdr
```

**Comment by jbondhus** [comment]
Thank you so much @snibgo, you've been so helpful with this!😃 

I'll play around with this more tonight and get back to you.

**Comment by jbondhus** [comment]
Ok, so I ran on this image using the following command as suggested. Minor change to set exposure.

[attachment](https://jbondhus-sharing.s3.us-east-1.amazonaws.com/2022/2022-11-30/attachment.7z)

```
magick.exe ".\ProPhoto RGB\_1044023.tif" -evaluate Pow 1.8 -set colorspace RGB -red-primary 0.734699,0.265301 -green-primary 0.159597,0.840403 -blue-primary 0.036598,0.000105 -white-point 0.345704,0.358540 -set hdr:exposure 0.1 x3.hdr
```

The result is what shows on the right in the attachment, on an HDR display. This is accurate as far as the color difference and level of desaturation and tinting on the right. Obviously a symptom of color management issues.

The interesting bit, is the software I use to view the HDR image says that there's no profile. Photoshop displays the .hdr file exactly the same. It might be broken in both, only supporting sRGB... The fact that if I try to change the color profile in Photoshop and then re-open it and it still doesn't display lends credence to this idea.

**Comment by snibgo** [comment]
The first few bytes of your x3.hdr are:
```
#?RADIANCE\n
EXPOSURE=0.1\n
GAMMA=1\n
PRIMARIES=0.734699 0.265301 0.159597 0.840403 0.036598 0.000105 0.345704 0.35854\n
FORMAT=32-bit_rle_rgbe\n\n
-Y 2449 +X 3265\n
```
... where "\n" represents a single byte, the usual unix new-line.

That looks reasonable to me. Gimp can read x3.hdr, and the result looks dark (on my SDR system), but I don't know if that is reasonable.

**Comment by jbondhus** [comment]
Yeah, the more I look at this I'm wondering if it's just that *both* programs don't understand different color spaces in either of the HDR image formats I've tried with.

I might try JPEG XR, though I think I'll have to compile "JXREncApp.exe" from scratch, I'm having no luck finding a binary for it, seems the windows version used to be hosted on codeplex and never got migrated over. Anyone happen to have a binary of that helper app? ImageMagick calls it.

![image](https://user-images.githubusercontent.com/1085350/204843780-5cb5077e-7eed-4d37-be37-9f04d994dd3a.png)

**Comment by jbondhus** [comment]
Yeah, the colors of that .hdr file are wrong in gimp as well... So so far we're 0 for 3. Not sure if that means there's still something we're missing in the metadata, or a problem with all three viewers.

See attached for a picture comparing the colors, I realized I forgot to include it in my earlier comments' attachment. The left is how the colors should look of the original TIFF in photoshop, the right is the HDR viewing program (though the .hdr file looks like this in photoshop and gimp (with darker colors) as well. Gimp might not be handling the exposure correctly, that might be why it's so dark - the colors look identical to photoshop and the HDR viewer when you boost the exposure.

![_1001695](https://user-images.githubusercontent.com/1085350/204846547-8b100ca0-1b2b-4fa5-8dfe-44714005a015.jpg)

**Reply by fmw42** [reply]
I don't know if this is relevant, but Photoshop allows for spot colors. Imagemagick does not use them. Can you post your original PS TIFF file, if you have not already.

**Reply by jbondhus** [reply]
Here's a link to it again.

https://jbondhus-sharing.s3.us-east-1.amazonaws.com/2022/2022-11-30/attachment.7z

**Reply by fmw42** [reply]
Again, I am a bit out of the loop here, but I note that Photoshop says the file has an embedded profile. Are you using that?

![Screen Shot 2022-11-30 at 9 37 16 AM](https://user-images.githubusercontent.com/13458218/204869253-6fa1fd15-d180-4b38-9f4d-ba93a1f51f85.png)

Second Photoshop as a view uses dot gain, which imagemagick and likely many other viewer do not use. So it may not be fair to compare if both images are not viewed by the same viewer and processed to make use of the dot gain.

![Screen Shot 2022-11-30 at 9 38 23 AM](https://user-images.githubusercontent.com/13458218/204869490-d2a0c78c-9033-40f9-938f-0fffb366a1b9.png)

**Comment by jbondhus** [comment]
Just an update, I've posted an issue on the project for the HDR photo viewer, listed below, and referenced this discussion in that. I'd like to find out if this is a problem with the file metadata, or an issue with the viewer.

https://github.com/13thsymphony/HDRImageViewer/issues/54

**Comment by snibgo** [comment]
I have identified and reported a bug in IM. See https://github.com/ImageMagick/ImageMagick/issues/5835 . Effectively, IM ignores chromaticities when *reading* a .HDR file. Any software that uses IM to read .HDR files will inherit this problem.

Regarding Gimp: I don't know if Gimp uses IM to read HDR files. If I read x3.hdr in Gimp, Gimp says the colorspace is "GIMP built-in Linear sRGB". I don't know if Gimp has *converted* the image to that colorspace, or simply *assigned* it.

In Gimp, we can *assign* a linear ProPhotoRGB profile (using Elle's profile I mentioned above). Then apply a simple linear transformation to lighten the image. The result on my SDR computer looks to me very similar to "The left is how the colors should look of the original TIFF in photoshop".

![clp](https://user-images.githubusercontent.com/33812368/204876935-48dc15d6-027a-4a4d-8b45-538bcb523150.jpg)
